Articles plagiarism check implementation

kol3x opened the pull request:
Closes #424 

Hi, I am starting this as draft request for now, because I still need to test the GitHub workflow script properly. I've never worked with one before, so it's ruff. The one I have now is work in progress.

However, the plagiarism-checker service is ready and working. Since Evgeny gave me a hint that this problem has a better solution than just using a paid third-party API, I started researching on how I might implement it otherwise. One old open-source [library](https://github.com/architshukla/Plagiarism-Checker) drew my attention, as it was using google search API to check text for plagiarism in the open web. I found that method better for given problem, but I didn't want to use a library that hasn't been updated for 10+ years, so I developed a little service that uses the same technic.

Additional costs with this method should be 0, since Google [lets you](https://developers.google.com/custom-search/v1/overview#pricing) make up to 100 queries a day for free. My service is implemented in a way, that it checks every sentence for a 100% match on the internet, while **Crypto Attack Wiki** article has less than 30 sentences on the average (Rough estimate). And since new articles are released not that often, we can have 2-3 free checks for a new article in a single day.

kol3x commented:
I finished and tested the whole thing. This is the end results for different text inputs: 
![image](https://github.com/1712n/dn-institute/assets/45823609/2ceea3b6-e5bc-463a-99ba-b0018577179d)
![image](https://github.com/1712n/dn-institute/assets/45823609/58a4928c-b421-4bab-b66d-dcb107c72605)

A little more on the implementation:

+ GitHub Actions workflow finds the first .md file in the PR and sends it to a Cloudflare worker. Then it formats the result to  a proper markdown for a GitHub PR comment. Lastly, it posts the results comment. There is one little hack I used here - for line breaks I literally pressed "Enter" in the .yml file, because neither \n or < b r > did the job. (\n didn't do anything at all, and < b r >  was creating a new line visually, but was still treating the whole comment as a single line, so for example if I had triple hash for my first line, everything in the comment got treated like a header)

+ Cloudflare worker checks the text one sentence at a time (Anything more than 10 characters and ending with a dot is considered a sentence) Now, this is not a super advanced way for checking plagiarism, but since Evgeny pointed out that we are looking for the simple copy-pasting of text from the open web - it does the job. 

I am gonna work on a README later with instructions on setting up the service and adding all necessary secrets

Looking forward to hear some feedback on my implementation!

kol3x commented:
@Lavriz review this please, whenever you are available

Lavriz commented on tools/plagiarism-checker/src/index.ts:
```diff
@@ -0,0 +1,93 @@
+
+  const plagiarismResults: PlagiarismResults[] = []
+
+  await Promise.all(
+    sentences.map(async (sentence) => {
+      const query = encodeURIComponent(sentence.trim())
+      const url = `https://www.googleapis.com/customsearch/v1?exactTerms=${query}&key=${c.env.GOOGLE_API_KEY}&cx=${c.env.GOOGLE_SEARCH_ENGINE_CX}&num=1`
+      const response = await fetch(url)
+      const searchResults: SearchResponse = await response.json()
+
+      if (!searchResults.items) {
+        return
+      }
+      plagiarismResults.push({
+        sentence,
+        matches: searchResults.items.map((item: SearchResult) => ({
+          title: item.title,
+          link: item.link
+        }))
+      })
+    })
+  )
```
Modifying a shared resource like `plagiarismResults` within multiple asynchronous operations can sometimes lead to race conditions.

kol3x replied:
I changed it, now we are not pushing to an external array, but returning from Promise.all directly to a variable. Then we are filtering out null values from the array.

Lavriz commented on tools/plagiarism-checker/src/index.ts:
```diff
@@ -0,0 +1,93 @@
+    return c.json({ error: "Invalid JSON format" }, 400)
+  }
+
+  const sentences = text
+    .split(".")
+    // if it's less than 10 characters - we don't treat it as a sentence and skip it
+    .filter((sentence) => sentence.trim().length > 10)
```
1. What about egde cases where a sentence inside contains abbreviations, like `U.S`, `e.g.`, `etc.` ? Maybe consider an alternative approach to splitting.
2. How did you come up with a 10-character limit?

kol3x replied:
> 1. What about egde cases where a sentence inside contains abbreviations, like `U.S`, `e.g.` ? Maybe consider an alternative approach to splitting.
> 
>     2. How did you come up with a 10-character limit?

Basically 10-character limit is meant to filter out abbreviations and irrelevant pieces of text. We assume that anything less then 10 characters can't be properly checked for a plagiarism with our method, because it's probably 1-2 words and google will surely get some matches, even if the text is actually unique. I can see how that's not ideal.

Another way would be to take in every N words instead of a sentence. But if 1 sentence is copy-pasted from the web and the next one is not - they are gonna get mixed up and plagiarism will not be found (since we are looking for 100% matches with google api).

We could use a library for splitting sentences, but I found this cool [regex](https://stackoverflow.com/a/72280712/10957343) that I am going to use instead.  

Now another problem I found in the process is that it treats headers as part of the following sentence: (so did my split(".") method)
![image](https://github.com/1712n/dn-institute/assets/45823609/d081579d-d9f0-4876-9ea7-cfcd04ee6b44) 
To fix that, I added another regex to filter out headers. 

And I removed the 10-char min limit.

Lavriz commented on tools/plagiarism-checker/src/index.ts:
```diff
@@ -0,0 +1,93 @@
+    }, 0)
+  }
+
+  const plagiarismPercent = plagiarisedSentencesNum
+    ? (sentences.length / plagiarisedSentencesNum) * 100
+    : 0
```
Please check the formula.

kol3x replied:
> Please check the formula.

It might look a bit confusing, but I think it's correct. 

We don't want to divide by 0, so on the first line we are checking if plagiarisedSentencesNum is falsy (0). 

If it's not, we are dividing amount of sentences by amount of plagiarized sentences and multiplying by 100 to turn into % form.

If it's falsy - the percent is set to 0.

Also, there might be an edge-case where sentences length is zero, but I handled that earlier in the code in my latest commits

kol3x replied:
Wait, I got it... Has to be plagiarisedSentencesNum / sentences.length instead.
Fixed it. My bad :)

kol3x commented:
Hi, @Lavriz 

I resolved all the requested changes and added instructions on how to integrate the service and set up corresponding secrets.

Hope to hear back from you soon

kol3x commented:
- Moved formatting logic to javascript, fixed some formatting bugs along the way. The plagiarism result comments are much prettier and consistent now. 
- The service is called by /plagiarismcheck command
- Added permission check, so only users in WIKI_REVIEWERS can call the command
